url3426a97ee4_rtreffer_minidns_tgz-pJ8-de_measite_minidns_minidnsrepl_MiniDnsReplJ8/src/de/measite/minidns/dnssec/DNSSECClient.java

```diff
@@ -0,0 +1,555 @@
+            } else {
+                toBeVerified.removeAll(records);
+            }
+            toBeVerified.remove(sigRecord);
+        }
+        return result;
+    }
+
+    private static boolean isParentOrSelf(String child, String parent) {
+        if (child.equals(parent)) return true;
+        if (parent.isEmpty()) return true;
+        String[] childSplit = child.split("\\.");
+        String[] parentSplit = parent.split("\\.");
+        if (parentSplit.length > childSplit.length) return false;
+        for (int i = 1; i <= parentSplit.length; i++) {
+            if (!parentSplit[parentSplit.length - i].equals(childSplit[childSplit.length - i])) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private Set<UnverifiedReason> verifySignedRecords(Question q, RRSIG rrsig, List<Record<? extends Data>> records) throws IOException {
+        Set<UnverifiedReason> result = new HashSet<>();
+        DNSKEY dnskey = null;
+        if (rrsig.typeCovered == TYPE.DNSKEY) {
+            // Key must be present
+            for (Record<? extends Data> record : records) {
+                Record<DNSKEY> dnsKeyRecord = record.ifPossibleAs(DNSKEY.class);
+                if (dnsKeyRecord == null) continue;
+
+                if (dnsKeyRecord.payloadData.getKeyTag() == rrsig.keyTag) {
+                    dnskey = dnsKeyRecord.payloadData;
+                    break;
+                }
+            }
+        } else if (q.type == TYPE.DS && rrsig.signerName.equals(q.name)) {
+            // We should not probe for the self signed DS negative response, as it will be an endless loop.
+            result.add(new NoTrustAnchorReason(q.name.ace));
+            return result;
+        } else {
+            DNSSECMessage dnskeyRes = queryDnssec(rrsig.signerName, TYPE.DNSKEY);
+            if (dnskeyRes == null) {
+                throw new DNSSECValidationFailedException(q, "There is no DNSKEY " + rrsig.signerName + ", but it is used");
+            }
+            result.addAll(dnskeyRes.getUnverifiedReasons());
+            for (Record<? extends Data> record : dnskeyRes.answerSection) {
+                Record<DNSKEY> dnsKeyRecord = record.ifPossibleAs(DNSKEY.class);
+                if (dnsKeyRecord == null) continue;
+
+                if (dnsKeyRecord.payloadData.getKeyTag() == rrsig.keyTag) {
+                    dnskey = dnsKeyRecord.payloadData;
+                }
+            }
+        }
+        if (dnskey == null) {
+            throw new DNSSECValidationFailedException(q, records.size() + " " + rrsig.typeCovered + " record(s) are signed using an unknown key.");
+        }
+        UnverifiedReason unverifiedReason = verifier.verify(records, rrsig, dnskey);
+        if (unverifiedReason != null) {
+            result.add(unverifiedReason);
+        }
+        return result;
+    }
+
+    private Set<UnverifiedReason> verifySecureEntryPoint(Question q, final Record<DNSKEY> sepRecord) throws IOException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 19. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 385-392 into a separate method.